sch_yang_condition.c

```diff
@@ -0,0 +1,923 @@
+} yang_result;
+
+typedef struct _yang_exists
+{
+    bool match;
+    char *path;
+} yang_exists;
+
+typedef enum
+{
+    PROC_F_IF_FEATURE = (1 << 0),  /* Parent is a if-feature */
+} proc_flags;
+
+
+static void sch_yang_process_xnode (sch_instance *instance, GNode *root, char *path,
+                                    char *step_path, xpath_node *xnode,
+                                    yang_result *presult, int depth, int *flags);
+
+
+static void
+sch_axis_child (char *path, char *child_name, yang_result *presult, int flags)
+{
+    if (child_name)
+    {
+        if ((flags & PROC_F_IF_FEATURE))
+            presult->value = g_strdup (child_name);
+        else
+            presult->value = g_strdup_printf ("%s/%s", path, child_name);
+
+        presult->result = true;
+    }
+    else
+    {
+        presult->result = false;
+    }
+}
+
+static void
+sch_axis_parent (char *path, char *step_path, yang_result *presult, bool self, int flags)
+{
+    char **split;
+    int count;
+    bool same = (!step_path || g_strcmp0 (path, step_path) == 0);
+
+    presult->result = false;
+    if (self && step_path && !same)
+    {
+        presult->result = true;
+        presult->value = g_strdup (step_path);
+        return;
+    }
+
+    if (step_path)
+        split = g_strsplit (step_path, "/", -1);
+    else
+        split = g_strsplit (path, "/", -1);
+
+    if (split)
+    {
+        count = g_strv_length (split);
+        if (count > 1)
+        {
+            g_free (split[count - 1]);
+            split[count - 1] = NULL;
+            presult->value = g_strjoinv ("/", split);
+            presult->result = true;
+        }
+        g_strfreev (split);
+    }
+}
+
+static gboolean
+sch_yang_step_exists_traverse_nodes (GNode *node, gpointer data)
+{
+    yang_exists *exists = (yang_exists *) data;
+    bool ret = false;
+    char *path = apteryx_node_path (node);
+
+    if (g_strcmp0 (path, exists->path) == 0)
+    {
+        exists->match = true;
+        ret = true;
+    }
+
+    g_free (path);
+    return ret;
+}
+
+static void
+sch_yang_step_exists (GNode *root, char *path, yang_result *presult)
+{
+    yang_exists exists = { };
+
+    exists.path = path;
+    g_node_traverse (root, G_IN_ORDER, G_TRAVERSE_ALL, -1,
+                     sch_yang_step_exists_traverse_nodes, &exists);
+    presult->result = exists.match;
+    if (!presult->result)
+    {
+        GNode *tree = apteryx_get_tree (path);
+
+        if (tree)
+        {
+            presult->result = true;
+            apteryx_free_tree (tree);
+        }
+    }
+}
+
+static void
+sch_yang_process_operator (sch_instance *instance, GNode *root, char *path,
+                           char *step_path, xpath_node *xnode, yang_result *presult,
+                           int depth, int *flags)
+{
+    char *new_step_path = NULL;
+    bool is_number = false;
+    bool error = false;
+    yang_result lresult = { };
+    yang_result rresult = { };
+    int lnumber = 0;
```

Comment:
I think we might need to be careful about sizes of integers here.
Also, from memory, strtol return 0 if there is an error in the string, which concerns me a bit. Never sure what is the best routine to call for character to integer conversion.

Reply:
I changed the in to a int64_t to allow for most types of ints.